### ISSUE TITLE
Client outside AWS fails to connect to cluster in AWS

### ISSUE THREAD

**Issue by mmitya** [hazelcast/hazelcast-aws#59]
Hello.

I have: 
- hazelcast 3.9.2 cluster is in AWS 
- hazelcast 3.9.3 client + hazelcast-aws 2.1.0 is outside AWS 

Client fails to connect: `Unable to retrieve credentials from IAM Role.`

Client log:
```2018-03-14 13:30:19,158 INFO [main] com.hazelcast.core.LifecycleService - hz.client_0 [PrometheusDEV] [3.9.3] HazelcastClient 3.9.3 (20180216 - 539b124) is STARTING 
2018-03-14 13:30:19,445 INFO [main] com.hazelcast.aws.AwsDiscoveryStrategy - Describe instances will be queried with iam-role, please make sure given iam-role have ec2:DescribeInstances policy attached. 
2018-03-14 13:30:19,529 INFO [main] com.hazelcast.core.LifecycleService - hz.client_0 [PrometheusDEV] [3.9.3] HazelcastClient 3.9.3 (20180216 - 539b124) is STARTED 
2018-03-14 13:30:19,554 DEBUG [main] com.hazelcast.internal.networking.nio.NioEventLoopGroup - hz.client_0 [PrometheusDEV] [3.9.3] TcpIpConnectionManager configured with Non Blocking IO-threading model: 1 input threads and 1 output threads 
2018-03-14 13:30:19,555 DEBUG [main] com.hazelcast.internal.networking.nio.NioEventLoopGroup - hz.client_0 [PrometheusDEV] [3.9.3] IO threads selector mode is SELECT 
2018-03-14 13:30:19,775 WARN [hz.client_0.cluster-] com.hazelcast.client.connection.ClientConnectionManager - hz.client_0 [PrometheusDEV] [3.9.3] Could not connect to cluster, shutting down the client. Unable to retrieve credentials from IAM Role: prometheus-hazelcast-core-standalone 
2018-03-14 13:30:19,776 INFO [hz.client_0.clientShutdown-] com.hazelcast.core.LifecycleService - hz.client_0 [PrometheusDEV] [3.9.3] HazelcastClient 3.9.3 (20180216 - 539b124) is SHUTTING_DOWN 
2018-03-14 13:30:19,778 INFO [hz.client_0.clientShutdown-] com.hazelcast.core.LifecycleService - hz.client_0 [PrometheusDEV] [3.9.3] HazelcastClient 3.9.3 (20180216 - 539b124) is SHUTDOWN
```

Client exception:
```Exception in thread "main" com.hazelcast.config.InvalidConfigurationException: Unable to retrieve credentials from IAM Role: prometheus-hazelcast-core-standalone 
at com.hazelcast.aws.impl.DescribeInstances.fillKeysFromIamRole(DescribeInstances.java:105) 
at com.hazelcast.aws.impl.DescribeInstances.fillKeysFromIamRoles(DescribeInstances.java:85) 
at com.hazelcast.aws.impl.DescribeInstances.execute(DescribeInstances.java:219) 
at com.hazelcast.aws.AWSClient.getAddresses(AWSClient.java:56) 
at com.hazelcast.aws.AwsDiscoveryStrategy.discoverNodes(AwsDiscoveryStrategy.java:146) 
at com.hazelcast.spi.discovery.impl.DefaultDiscoveryService.discoverNodes(DefaultDiscoveryService.java:74) 
at com.hazelcast.client.spi.impl.discovery.DiscoveryAddressProvider.loadAddresses(DiscoveryAddressProvider.java:44) 
at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl.getPossibleMemberAddresses(ClientConnectionManagerImpl.java:895) 
at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl.connectToClusterInternal(ClientConnectionManagerImpl.java:812) 
at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl.access$1200(ClientConnectionManagerImpl.java:103) 
at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl$2.call(ClientConnectionManagerImpl.java:859) 
at com.hazelcast.client.connection.nio.ClientConnectionManagerImpl$2.call(ClientConnectionManagerImpl.java:855) 
at java.util.concurrent.FutureTask.run(FutureTask.java:266) 
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) 
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) 
at java.lang.Thread.run(Thread.java:748) 
at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:64) 
at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:80) 
Caused by: com.hazelcast.config.InvalidConfigurationException: Unable to lookup role in URI: http://169.254.169.254/latest/meta-data/iam/security-credentials/prometheus-hazelcast-core-standalone 
at com.hazelcast.aws.utility.MetadataUtil.retrieveMetadataFromURI(MetadataUtil.java:75) 
at com.hazelcast.aws.impl.DescribeInstances.retrieveRoleFromURI(DescribeInstances.java:136) 
at com.hazelcast.aws.impl.DescribeInstances.fillKeysFromIamRole(DescribeInstances.java:101) 
... 17 more 
Caused by: java.net.SocketException: Network is unreachable: connect 
at java.net.DualStackPlainSocketImpl.waitForConnect(Native Method) 
at java.net.DualStackPlainSocketImpl.socketConnect(DualStackPlainSocketImpl.java:85) 
at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350) 
at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206) 
at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188) 
at java.net.PlainSocketImpl.connect(PlainSocketImpl.java:172) 
at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) 
at java.net.Socket.connect(Socket.java:589) 
at sun.net.NetworkClient.doConnect(NetworkClient.java:175) 
at sun.net.www.http.HttpClient.openServer(HttpClient.java:463) 
at sun.net.www.http.HttpClient.openServer(HttpClient.java:558) 
at sun.net.www.http.HttpClient.<init>(HttpClient.java:242) 
at sun.net.www.http.HttpClient.New(HttpClient.java:339) 
at sun.net.www.http.HttpClient.New(HttpClient.java:357) 
at sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:1220) 
at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1156) 
at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1050) 
at sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:984) 
at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1564) 
at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492) 
at com.hazelcast.aws.utility.MetadataUtil.retrieveMetadataFromURI(MetadataUtil.java:67) 
... 19 more
```

Test client code:
```ClientConfig clientConfig = new ClientConfig(); 
clientConfig.getProperties().setProperty("hazelcast.discovery.enabled", "true"); 
clientConfig.getProperties().setProperty("hazelcast.discovery.public.ip.enabled", "true"); 
Map<String, Comparable> properties = new HashMap<>(); 
properties.put("iam-role", "prometheus-hazelcast-core-standalone"); 
properties.put("region", "eu-west-1"); 
properties.put("host-header", "ec2.amazonaws.com"); 
properties.put( 
"security-group-name", 
"security-group-name"); 
properties.put("tag-key", "role_name"); 
properties.put("tag-value", "prometheus-hazelcast-core-standalone"); 
properties.put("hzPort", "5701"); 
AwsDiscoveryStrategyFactory awsDiscoveryStrategyFactory = new AwsDiscoveryStrategyFactory(); 
DiscoveryStrategyConfig discoveryStrategyConfig = new DiscoveryStrategyConfig(awsDiscoveryStrategyFactory, 
properties); 
ClientNetworkConfig clientNetworkConfig = clientConfig.getNetworkConfig(); 
clientNetworkConfig.getDiscoveryConfig().addDiscoveryStrategyConfig(discoveryStrategyConfig); 
clientConfig.getGroupConfig().setName("PrometheusDEV"); 
clientConfig.getGroupConfig().setPassword("PrometheusDEV"); 
// 
HazelcastInstance client = HazelcastClient.newHazelcastClient(clientConfig);
```

Please assist.
Thanks.

**Comment by leszko** [comment]
Hi @mmitya ,

Currently  setting "name" and "password":

```
clientConfig.getGroupConfig().setName("PrometheusDEV"); 
clientConfig.getGroupConfig().setPassword("PrometheusDEV"); 
```

is not supported. So you may use IAM role feature only from inside AWS. 

Can't you generate Access Key and use it instead?

**Comment by mmitya** [comment]
Hi @leszko 
Will give it a try.
Thanks.

**Comment by gokhanoner** [comment]
@mmitya  `iam-role` can be used only when you're inside AWS. Otherwise, you need to use Access Key & Secure Key.

**Comment by mesutcelik** [comment]
@gokhanoner Do you think we can close this one by updating README with that information?

**Comment by gokhanoner** [comment]
@mesutcelik I believe so since AWS Service returning `iam-role` related info is only available within AWS.

**Comment by mesutcelik** [comment]
Ok. We will update this [comment](https://github.com/hazelcast/hazelcast-aws/blob/master/src/main/java/com/hazelcast/aws/utility/MetadataUtil.java#L36) and README first then close the issue. Thx!

**Comment by dbrimley** [comment]
Fixed by #67
